src/main/scala/com/itechart/ny_accidents/report/generators/ReportGenerators.scala

```diff
@@ -0,0 +1,98 @@
+    report.generateReportForTupleRDD(data, DAY_OF_WEEK_REPORT_HEADER)
+  }
+}
+
+class HourOfDayReportGenerator extends ReportGenerator {
+  override def calculateReport(data: RDD[MergedData]): RDD[_ <: Product] = {
+    TimeMetricService.countHours(data)
+  }
+
+  override def generateReport(data: RDD[_ <: Product], report: Reports): Seq[Seq[String]] = {
+    report.generateReportForTupleRDD(data, HOUR_OF_DAY_REPORT_HEADER)
+  }
+}
+
+class PeriodReportGenerator extends ReportGenerator {
+  override def calculateReport(data: RDD[MergedData]): RDD[_ <: Product] = {
+    WeatherMetricService.definePeriod(data)
+  }
+
+  override def generateReport(data: RDD[_ <: Product], report: Reports): Seq[Seq[String]] = {
+    report.generateReportForTupleRDD(data, DAY_OF_WEEK_REPORT_HEADER)
+  }
+}
+
+class WeatherReportGenerator extends ReportGenerator {
+  override def calculateReport(data: RDD[MergedData]): RDD[_ <: Product] = {
+    WeatherMetricService.getPhenomenonPercentage(data)
+  }
+
+  override def generateReport(data: RDD[_ <: Product], report: Reports): Seq[Seq[String]] = {
+    report.generateReportForTupleRDD(data, PHENOMENON_REPORT_HEADER)
+  }
+}
+
+class BoroughReportGenerator extends ReportGenerator {
+  override def calculateReport(data: RDD[MergedData]): RDD[_ <: Product] = {
+    DistrictMetricService.getBoroughPercentage(data)
+  }
+
+  override def generateReport(data: RDD[_ <: Product], report: Reports): Seq[Seq[String]] = {
+    report.generateReportForTupleRDD(data, BOROUGH_REPORT_HEADER)
+  }
+}
+
+class DistrictReportGenerator extends ReportGenerator {
+  override def calculateReport(data: RDD[MergedData]): RDD[_ <: Product] = {
+    DistrictMetricService.getDistrictsPercentage(data)
+  }
+
+  override def generateReport(data: RDD[_ <: Product], report: Reports): Seq[Seq[String]] = {
+    report.generateReportForTupleRDD(data, DISTRICT_REPORT_HEADER)
+  }
+}
+
+class PopulationToNumberOfAccidentsReportGenerator @Inject()(populationService: PopulationMetricService) extends ReportGenerator {
+  override def calculateReport(data: RDD[MergedData]): RDD[_ <: Product] = {
+    populationService.getPopulationToNumberOfAccidentsRatio(data)
+  }
+
+  override def generateReport(data: RDD[_ <: Product], report: Reports): Seq[Seq[String]] = {
+    report.generateReportForTupleRDD(data, DISTRICT_REPORT_HEADER)
+  }
+}
+
+class AccidentCountDuringPhenomenonPerHourReportGenerator extends ReportGenerator {
+  def calculateReport(data: RDD[MergedData], weatherPhenomenon: RDD[(String, Int, Double)]): RDD[_ <: Product] = {
+    WeatherMetricService.calculateAccidentCountDuringPhenomenonPerHour(data, weatherPhenomenon)
+  }
+
+  override def generateReport(data: RDD[_ <: Product], report: Reports): Seq[Seq[String]] = {
+    report.generateReportForTupleRDD(data, ACCIDENTS_DURING_PHENOMENON_COUNT_REPORT_HEADER)
+  }
+
+  override def calculateReport(data: RDD[MergedData]): RDD[_ <: Product] = {
+    Spark.sc.parallelize(Seq())
```

Comment:
What purpose of this method? It ignores parameter data and just returns empty RDD of Seq(). Please clarify

Reply:
I just didn’t think how to pass 2 arguments to the overridden method

Reply:
Because there is problem with code design. You say that _AccidentCountDuringPhenomenonPerHourReportGenerator_  is _ReportGenerator_ but in fact it's not as some methods are not needed Current implementation of _AccidentCountDuringPhenomenonPerHourReportGenerator_ will confused client as it will expect some Product will be generated from _data_ but in fact nothing will happen.

As a short solution that came into my head is What If we'd split _ReportGenerator_ into two traits: _ReportGenerator_ and _ReportCalculator_? What do you think?

Reply:
Ok, but. What to do with problem with different number of arguments?
For example, trait ReportCalculator contains method like `def calculate(data: RDD[MergedData])`.

And in _AccidentCountDuringPhenomenonPerHourReportGenerator_ we have to pass not only data.

Reply:
Discussed internally. Waiting for changes...